compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/FragmentLocalFunctionPatchLowering.kt

```diff
@@ -0,0 +1,112 @@
+        if (declaration.name.asString() != GENERATED_FUNCTION_NAME) return declaration
+
+        declaration.body!!.transformChildrenVoid(object : IrElementTransformerVoidWithContext() {
+            override fun visitCall(expression: IrCall): IrExpression {
+                expression.transformChildrenVoid(this)
+                val remappedTarget: LocalDeclarationsLowering.LocalFunctionContext =
+                    context.localFunctions[expression.symbol.owner] ?: return super.visitCall(expression)
+
+                val irBuilder = context.createJvmIrBuilder(declaration.symbol)
+                return irBuilder.irCall(remappedTarget.transformedDeclaration).apply {
+                    this.copyTypeArgumentsFrom(expression)
+                    extensionReceiver = expression.extensionReceiver
+                    dispatchReceiver = expression.dispatchReceiver
+
+                    remappedTarget.transformedDeclaration.valueParameters.map { newValueParameterDeclaration ->
+                        val oldParameter = context.newParameterToOld[newValueParameterDeclaration]
+
+                        val getValue = if (oldParameter != null) {
+                            // the parameter is an actual parameter to the local
+                            // function, not a parameter corresponding to a
+                            // capture introduced by a lowering: fetch
+                            // the corresponding argument from the existing
+                            // call and place at the appropriate slot in the
+                            // call to the lowered function
+                            expression.getValueArgument(oldParameter.index)!!
+                        } else {
+                            // The parameter is introduced by the lowering to
+                            // private static function, so corresponds to a _capture_ by the local function
+                            val capturedValueSymbol =
+                                context.newParameterToCaptured[newValueParameterDeclaration]
+                                    ?: error("Non-mapped parameter $newValueParameterDeclaration")
+
+                            // We introduce a new parameter to the _fragment function_ surrounding the call to the
+                            // lowered local function, and supply _that_ parameter to the corresponding _argument_ slot
+                            // in the call to the lowered function.
+                            val newParameter = declaration.addValueParameter {
+                                type = capturedValueSymbol.owner.type
+                                name = capturedValueSymbol.owner.name
+                            }
+
+                            context.state.recordNewFragmentCaptureParameter(
+                                newParameter.name.asString(),
+                                capturedValueSymbol.descriptor.type,
+                                capturedValueSymbol.descriptor
```

Comment:
Can we use IR-based descriptors here (and thus remove `@OptIn` on the function)?
```
                                capturedValueSymbol.owner.type.toIrBasedKotlinType(),
                                capturedValueSymbol.owner.toIrBasedDescriptor()
```

Reply:
Yes, that works!